account/account.go

```diff
@@ -120,3 +136,79 @@ func setDerivedConfig(account *config.Account) {
 		}
 	}
 }
+
+// PatchAccount represents the GDPR portion of a publisher account configuration that can be mutated
+// for backwards compatibility reasons
+type PatchAccount struct {
+	GDPR map[string]*PatchAccountGDPRPurpose `json:"gdpr"`
+}
+
+// PatchAccountGDPRPurpose represents account-specific GDPR purpose configuration data that can be mutated
+// for backwards compatibility reasons
+type PatchAccountGDPRPurpose struct {
+	EnforceAlgo    string `json:"enforce_algo,omitempty"`
+	EnforcePurpose *bool  `json:"enforce_purpose,omitempty"`
+}
+
+// ConvertGDPREnforcePurposeFields is responsible for ensuring account GDPR config backwards compatibility
+// given the recent type change of gdpr.purpose{1-10}.enforce_purpose from a string to a bool. This function
+// iterates over each GDPR purpose config and sets enforce_purpose and enforce_algo to the appropriate
+// bool and string values respectively if enforce_purpose is a string and enforce_algo is not set
+func ConvertGDPREnforcePurposeFields(config []byte) (newConfig []byte, err error) {
+	gdprJSON, _, _, err := jsonparser.Get(config, "gdpr")
+	if err != nil && err == jsonparser.KeyPathNotFoundError {
+		return config, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	newAccount := PatchAccount{
+		GDPR: map[string]*PatchAccountGDPRPurpose{},
+	}
+
+	for i := 1; i <= 10; i++ {
+		purposeName := fmt.Sprintf("purpose%d", i)
+
+		enforcePurpose, purposeDataType, _, err := jsonparser.Get(gdprJSON, purposeName, "enforce_purpose")
+		if err != nil && err == jsonparser.KeyPathNotFoundError {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		if purposeDataType != jsonparser.String {
+			continue
+		}
+
+		_, _, _, err = jsonparser.Get(gdprJSON, purposeName, "enforce_algo")
+		if err != nil && err != jsonparser.KeyPathNotFoundError {
+			return nil, err
+		}
+		if err == nil {
+			continue
+		}
+
+		newEnforcePurpose := false
+		if string(enforcePurpose) == "full" {
+			newEnforcePurpose = true
+		}
+
+		newAccount.GDPR[purposeName] = &PatchAccountGDPRPurpose{
+			EnforceAlgo:    "full",
+			EnforcePurpose: &newEnforcePurpose,
+		}
+	}
+
+	patchConfig, err := json.Marshal(newAccount)
+	if err != nil {
+		return nil, err
+	}
+
+	newConfig, err = jsonpatch.MergePatch(config, patchConfig)
+	if err != nil {
+		return nil, err
```

Comment:
Hm I don't see a way to test this. For the merge patch to fail either the config or the patch has to be malformed. A malformed config will never make it this far because an error will be caught at the top of the function when calling `jsonparser.Get(config, "gdpr")` which I already have a test case for. And the patch cannot be malformed because we are correctly programmatically constructing it here. The function would need to be refactored to achieve this and that probably isn't worth it.